### ISSUE TITLE
Statistical osu SR & performance calc 

### ISSUE THREAD

**Issue by joseph-ireland** [ppy/osu-performance#92]
I'm proposing a new SR & difficulty calculation which aims to:
- properly incorporate diff spikes and length into SR - SR now represents difficulty to FC, not peak difficulty.
- give better judgements of imperfect plays by assessing the difficulty to get a given combo and miss count, rather than a flat punishment for every map

This issue is the same as the bottom half of #64 which has become a bit hard to follow. 

There's a draft writeup of the methodology [here](https://docs.google.com/document/d/18Di9enQzy49VpGgtyH8RF6ljVF3Vx0Z9EzAgWJEhj7U/edit?usp=sharing).
The code is at PR X (I've also got an osu-tools branch [here](https://github.com/joseph-ireland/osu-tools/tree/prob_stars_combo_scaling)  with a few improvements)

A few results: 

[freddie benson](https://paste.ee/p/8SaGi)
[walkingtuna](https://paste.ee/p/q3BDp)
[idke](https://paste.ee/p/540sR)
[nathan on osu](https://paste.ee/p/nW2vf)
[karthy](https://paste.ee/p/3Z3T1)

[me (~50k)](https://paste.ee/p/VcJRE)
[starbin1 (~100k)](https://paste.ee/p/vgGrf)

Todo:

- [ ] tweak parameters for best results. The most important are StarBonusPerLengthDouble and StarBonusBaseTime in osu.Game.Rulesets.Osu/Difficulty/OsuSkill.cs (they can also be overriden separately in Aim.cs and Speed.cs)
- [ ] make abraker95's suggested changes to the write-up
- [ ] possibly refactor to simplify the implementation - it's currently transforming strain using log(x) before doing calculations, then transforming back at the end. Following the log(x) through all the calculations simplifies some stuff, adds more concrete meaning to the parameters, and makes the implementation match the writeup more closely.

**Comment by HeBuwei** [comment]
I feel the expected length is a perfect measure for pp but not suitable for SR.

Under the live SR algorithm, if you add an easy part at the beginning of a hard map, the SR basically won't change. This behavior makes perfect sense, and people are already used to it, so it would be good to preserve it. The expected length approach does not have this property.

I personally would choose FC probability per play for SR calculation. It has the aforementioned property, and the way it changes with length is similar to what we have now.

**Comment by tr3sleches** [comment]
https://github.com/ppy/osu-performance/issues/64#issuecomment-493814581
This is my response to your last statement in issue #64, just got around to responding.

**Comment by joseph-ireland** [comment]
@HeBuwei I agree, I thought this too, FC probability is the closest analogue to current star rating. I quite like the idea that star rating more closely reflects PP, but I guess that's not what it's for. I was undecided between which one to use, but it was easier to just use the same value as the PP calculations. I haven't actually looked at star ratings in detail yet, I was just focused on PP, so maybe that's something to compare.

**Comment by tr3sleches** [comment]
Here is your proof for the expected time to FC:

[Expected Time to FC proof.docx](https://github.com/ppy/osu-performance/files/3244157/Expected.Time.to.FC.proof.docx)


In hindsight, I shouldn't even have to do this because all this is a mathy way of saying you retry the map from the beginning every time you miss. 
I hate proofs.

**Comment by joseph-ireland** [comment]
I don't get it. Why number of retries and not number of attempts? The first attempt still takes time. Put in `p_i = 1` in your calculations and suddenly everything breaks. Change E[retries] to E[attempts] and it matches what I've got.

**Comment by tr3sleches** [comment]
If you get the section in 1 attempt, you don’t have to retry the map and the entire 
p_i = 1 doesn’t break it, pay attention. 
I formatted it this way for interpretation sake but that 1-p_k section cancels out so it still works.
The first attempt is different from the subsequent ones. If p_i = 1, R_i = 0
Infinity*0 is 0.
It doesn’t break...

**Comment by joseph-ireland** [comment]
OK, so R_i = 0. Putting that in your first equation gives `E[T_n] = delta t_{n-1}`, so you expect to FC the map in less time than one play through, which doesn't make sense. It should be  `E[T_n] = t_n`, i.e. `sum delta t_i` if all the p_i are 1

**Comment by joseph-ireland** [comment]
Unless you're actually trying to calculate expected time spent in failed attempts before an FC, which actually isn't a bad idea, i.e. E[T_fc] - map_length. That's always going to have a limit of zero as skill goes to infinity, so you don't need to special case long maps, and all the calculations will still work, just need to replace the `targe_FC_time` with `target_FC_time + map_length`.

**Comment by tr3sleches** [comment]
If R_i = 0 for all i, you just add up the deltaT so you just get the map length

**Comment by tr3sleches** [comment]
I reversed some of the steps: 
Hope this clears stuff up.
[Expected Time to FC proof.docx](https://github.com/ppy/osu-performance/files/3244316/Expected.Time.to.FC.proof.docx)

**Comment by tr3sleches** [comment]
Oh, I see what you mean. I had this fixed when I worked this out on paper, forgot to put it in the word doc.
One sec.

**Comment by joseph-ireland** [comment]
wait., does E[T] = sum E[T_i]? i was thinking E[T] = E[T_n], so it wasn't making sense
